File: sample-apps/go-sample-app/collection/random_metrics.go
package collection

import (
	"context"
	"fmt"
	"math/rand"
	"time"

	"go.opentelemetry.io/otel/metric"
	"go.opentelemetry.io/otel/metric/instrument"
)

var (
	threadCount int64 = 0
	threadsBool       = true
)

// randomMetricCollector contains all the random based metric instruments.
type randomMetricCollector struct {
	timeAlive     instrument.Int64Counter
	cpuUsage      instrument.Int64ObservableGauge
	totalHeapSize instrument.Int64ObservableUpDownCounter
	threadsActive instrument.Int64UpDownCounter
	meter         metric.Meter
}

// NewRandomMetricCollector returns a new type struct that holds and registers the 4 random based metric instruments used in the Go-Sample-App;
// HeapSize, ThreadsActive, TimeAlive, CpuUsage
func NewRandomMetricCollector(mp metric.MeterProvider) randomMetricCollector {
	rmc := randomMetricCollector{}
	rmc.meter = mp.Meter("github.com/aws-otel-commnunity/sample-apps/go-sample-app/collection")
	rmc.registerHeapSize()
	rmc.registerThreadsActive()
	rmc.registerTimeAlive()
	rmc.registerCpuUsage()
	return rmc
}

// registerTimeAlive registers a Synchronous Counter called TimeAlive.
func (rmc *randomMetricCollector) registerTimeAlive() {
	timeAliveMetric, err := rmc.meter.Int64Counter(
		timeAlive+testingId,
		instrument.WithDescription("Total amount of time that the application has been alive"),
		instrument.WithUnit("ms"),
	)
	if err != nil {
		fmt.Println(err)
	}
	rmc.timeAlive = timeAliveMetric
}

// registerCpuUsage registers an Asynchronous Gauge called CpuUsage.
func (rmc *randomMetricCollector) registerCpuUsage() {
	cpuUsageMetric, err := rmc.meter.Int64ObservableGauge(
		cpuUsage+testingId,
		instrument.WithDescription("Cpu usage percent"),
		instrument.WithUnit("1"),
	)
	if err != nil {
		fmt.Println(err)
	}
	rmc.cpuUsage = cpuUsageMetric

}

// registerHeapSize registers an Asynchronous UpDownCounter called HeapSize.
func (rmc *randomMetricCollector) registerHeapSize() {
	totalHeapSizeMetric, err := rmc.meter.Int64ObservableUpDownCounter(
		totalHeapSize+testingId,
		instrument.WithDescription("The current total heap size"),
		instrument.WithUnit("By"),
	)
	if err != nil {
		fmt.Println(err)
	}
	rmc.totalHeapSize = totalHeapSizeMetric

}

// registerThreadsActive registers a Synchronous UpDownCounter called ThreadsActive.
func (rmc *randomMetricCollector) registerThreadsActive() {
	threadsActiveMetric, err := rmc.meter.Int64UpDownCounter(
		threadsActive+testingId,
		instrument.WithUnit("1"),
		instrument.WithDescription("The total amount of threads active"),
	)
	if err != nil {
		fmt.Println(err)
	}
	rmc.threadsActive = threadsActiveMetric
}

// UpdateMetricsClient generates new metric values for Synchronous instruments every TimeInterval and
// Asynchronous instruments every CollectPeriod configured by the controller.
func (rmc *randomMetricCollector) RegisterMetricsClient(ctx context.Context, cfg Config) {
	go func() {
		for {
			rmc.updateTimeAlive(ctx, cfg)
			rmc.updateThreadsActive(ctx, cfg)
			time.Sleep(time.Second * time.Duration(cfg.TimeInterval))
		}
	}()
	rmc.updateCpuUsage(ctx, cfg)
	rmc.updateTotalHeapSize(ctx, cfg)
}

// updateTimeAlive updates TimeAlive by TimeAliveIncrementer increments.
func (rmc *randomMetricCollector) updateTimeAlive(ctx context.Context, cfg Config) {
	rmc.timeAlive.Add(ctx, cfg.TimeAliveIncrementer*1000, randomMetricCommonLabels...) // in millisconds
}

// updateCpuUsage updates CpuUsage by a value between 0 and CpuUsageUpperBound every SDK call.
func (rmc *randomMetricCollector) updateCpuUsage(ctx context.Context, cfg Config) {
	min := 0
	max := int(cfg.CpuUsageUpperBound)
	if _, err := rmc.meter.RegisterCallback(
		// SDK periodically calls this function to collect data.
		func(ctx context.Context, o metric.Observer) error {
			cpuUsage := int64(rand.Intn(max-min) + min)
			o.ObserveInt64(rmc.cpuUsage, cpuUsage, randomMetricCommonLabels...)

			return nil
		},
		rmc.cpuUsage,
	); err != nil {
		panic(err)
	}
}

// updateTotalHeapSize updates HeapSize by a value between 0 and TotalHeapSizeUpperBound every SDK call.
func (rmc *randomMetricCollector) updateTotalHeapSize(ctx context.Context, cfg Config) {
	min := 0
	max := int(cfg.TotalHeapSizeUpperBound)
	if _, err := rmc.meter.RegisterCallback(
		// SDK periodically calls this function to collect data.
		func(ctx context.Context, o metric.Observer) error {
			totalHeapSize := int64(rand.Intn(max-min) + min)
			o.ObserveInt64(rmc.totalHeapSize, totalHeapSize, randomMetricCommonLabels...)

			return nil
		},
		rmc.totalHeapSize,
	); err != nil {
		panic(err)
	}
}

// updateThreadsActive updates ThreadsActive by a value between 0 and 10 in increments or decrements of 1 based on previous value.
func (rmc *randomMetricCollector) updateThreadsActive(ctx context.Context, cfg Config) {
	if threadsBool {
		if threadCount < int64(cfg.ThreadsActiveUpperBound) {
			rmc.threadsActive.Add(ctx, 1, randomMetricCommonLabels...)
			threadCount++
		} else {
			threadsBool = false
			threadCount--
		}

	} else {
		if threadCount > 0 {
			rmc.threadsActive.Add(ctx, -1, randomMetricCommonLabels...)
			threadCount--
		} else {
			threadsBool = true
			threadCount++
		}
	}
}

File: sample-apps/python-sample-app/random_metrics.py
import random
import threading
import time

from config import *
from opentelemetry.metrics import CallbackOptions, Observation
from opentelemetry import metrics

cfg = create_config('config.yaml')

meter = metrics.get_meter(__name__)

# Starts the callback for cpu usage
def cpu_usage_callback(options: CallbackOptions):
    min = 0
    max = cfg['random_cpu_usage_upper_bound']
    cpu_usage = Observation(value=random.randint(min, max))
    print('CPU Usage asked by SDK')
    yield cpu_usage

# Starts the callback for heap size
def heap_size_callback(options: CallbackOptions):
    min = 0
    max = cfg['random_total_heap_size_upper_bound']
    total_heap_size = Observation(value=random.randint(min, max))
    print("Heapsize asked by SDK")
    yield total_heap_size

# This is the random metric collector class
class random_metric_collector():
    
    # Init registers 4 different metrics
    def __init__(self):
        self.time_alive=meter.create_counter(
            name="time_alive",
            description="Increments by 1s every second the app is alive",
            unit='s'
        )
        self.cpu_usage=meter.create_observable_gauge(
            name="cpu_usage",
            callbacks=[cpu_usage_callback],
            description="The percentage of CPU usage in use",
            unit='1'
        )
        self.threads_active=meter.create_up_down_counter(
            name="threads_active",
            description="the total number of threads active",
            unit='1'
        )
        self.total_heap_size=meter.create_observable_gauge(
            name="total_heap_size",
            callbacks=[heap_size_callback],
            description="the current total heap size",
            unit='1'
        )
        self.thread_bool = True
        self.thread_count = 0

    # Adds one to the time alive counter
    def update_time_alive(self, cfg=None):
        self.time_alive.add(cfg['random_time_alive_incrementer'])
    
    # Updates the currently active threads based on its current bounds.
    def update_threads_active(self, cfg=None):
        if self.thread_bool:
            if self.thread_count < cfg['random_threads_active_upper_bound']:
                self.threads_active.add(1)
                self.thread_count += 1
            else:
                self.thread_bool = False
                self.thread_count -= 1
        
        else:
            if self.thread_count > 0 :
                self.threads_active.add(-1)
                self.thread_count -= 1
            else:
                self.thread_bool = True
                self.thread_count += 1

    # This function registers the metrics client 
    def register_metrics_client(self,cfg=None):
        
        def update(self, cfg):
            while True:
                self.update_time_alive(cfg)
                self.update_threads_active(cfg)
                print("updating time alive & active threads...")
                time.sleep(cfg['time_interval'])

        cpu_usage_callback_thread = threading.Thread(target=cpu_usage_callback, args=(cfg,) ,daemon=True)
        heap_size_callback_thread = threading.Thread(target=heap_size_callback, args=(cfg,), daemon=True)
        update_thread = threading.Thread(target=update, args=(self, cfg,), daemon=True)
        cpu_usage_callback_thread.start()
        heap_size_callback_thread.start()
        update_thread.start()

# rmc = random_metric_collector()
# rmc.register_metrics_client(cfg)